File: string/121.cpp
#include <bits/stdc++.h>

using namespace std;

void dfs(int start, string s, unordered_set<string> &uniqueSubs, int &maxQuantity) {
    if (start == s.size()) {
        maxQuantity = max(maxQuantity, static_cast<int>(uniqueSubs.size()));
        return;
    }

    for (int i = 1; i <= s.size(); ++i) {
        string chunk = s.substr(start, i);
        if (uniqueSubs.count(chunk) == 0) {
            uniqueSubs.insert(chunk);
            dfs(start + chunk.size(), s, uniqueSubs, maxQuantity);
            uniqueSubs.erase(chunk);
        }
    }
}

int maxUniqueSplit(string s) {
    unordered_set<string> uniqueSubs;
    int maxQuantity = -1;

    dfs(0, s, uniqueSubs, maxQuantity);

    return maxQuantity;
}

int main() {
    string s = "ababccc";
    cout << maxUniqueSplit(s);

    return 0;
}

File: string/121.py
def dfs(s, start, max_quantity, unique_subs):
    if start >= len(s):
        max_quantity[0] = max(max_quantity[0], len(unique_subs))
        return

    for i in range(start, len(s)):
        chunk = s[start : i + 1]

        if chunk in unique_subs:
            continue

        unique_subs.add(chunk)
        dfs(s, i + 1, max_quantity, unique_subs)
        unique_subs.remove(chunk)


def maxUniqueSplit(s: str) -> int:
    unique_subs = set()
    max_quantity = [0]

    dfs(s, 0, max_quantity, unique_subs)

    return max_quantity[0]


s = "a"
print(maxUniqueSplit(s))
